libs/beacon/src/beacon_manager.cpp

```diff
@@ -17,167 +17,617 @@
 //------------------------------------------------------------------------------
 
 #include "beacon/beacon_manager.hpp"
+#include "crypto/ecdsa.hpp"
 
 namespace fetch {
 namespace dkg {
 
-void BeaconManager::Reset(uint64_t cabinet_size, uint32_t threshold)
+bn::G2 BeaconManager::zeroG2_;
+bn::Fr BeaconManager::zeroFr_;
+bn::G2 BeaconManager::group_g_;
+bn::G2 BeaconManager::group_h_;
+
+constexpr char const *LOGGING_NAME = "DKG";
+
+BeaconManager::BeaconManager(CertificatePtr certificate)
+  : certificate_{std::move(certificate)}
 {
-  cabinet_size_ = cabinet_size;
-  threshold_    = threshold;
+  if (certificate_ == nullptr)
+  {
+    auto ptr = new crypto::ECDSASigner();
+    certificate_.reset(ptr);
+    ptr->GenerateKeys();
+  }
 
-  received_shares_.clear();
-  participants_.clear();
+  static std::once_flag flag;
+
+  std::call_once(flag, []() {
+    bn::initPairing();
```

Comment:
If this is here it should not be in block coordinator. Also, I might consider a pattern like the BLSInitialiser - that way it hides this away

Reply:
I have removed it from block coordinator.